independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ContextInstancesGenerator.java

```diff
@@ -182,42 +208,49 @@ private void implementRemoveEach(ClassCreator contextInstances, Map<String, Inst
         removeEach.returnVoid();
     }
 
-    private void implementRemove(ClassCreator contextInstances, List<BeanInfo> beans,
-            Map<String, InstanceAndLock> idToFields) {
+    private List<MethodDescriptor> implementRemove(ClassCreator contextInstances, List<BeanInfo> beans,
+            Map<String, InstanceAndLock> idToFields, Map<String, MethodDescriptor> lazyLocks) {
         MethodCreator remove = contextInstances
                 .getMethodCreator("remove", ContextInstanceHandle.class, String.class)
                 .setModifiers(ACC_PUBLIC);
 
         StringSwitch strSwitch = remove.stringSwitch(remove.getMethodParam(0));
         // https://github.com/quarkusio/gizmo/issues/164
         strSwitch.fallThrough();
+        List<MethodDescriptor> removeMethods = new ArrayList<>(beans.size());
         for (BeanInfo bean : beans) {
             InstanceAndLock fields = idToFields.get(bean.getIdentifier());
             FieldDescriptor instanceField = fields.instance;
             // There is a separate remove method for every instance handle field
             // To eliminate large stack map table in the bytecode
             MethodCreator removeHandle = contextInstances.getMethodCreator("r" + instanceField.getName(),
                     ContextInstanceHandle.class).setModifiers(ACC_PRIVATE);
-            // this.1l.lock();
             // ContextInstanceHandle<?> copy = this.1;
-            // if (copy != null) {
-            //    this.1 = null;
+            // if (copy == null) {
+            //  return null;
             // }
-            // this.1l.unlock();
+            // Lock lock = lazy1l();
+            // lock.lock();
+            // copy = this.1;
```

Comment:
I think that we should either skip the read operation `copy = this.1;` or replace `this.1 = null;` with `if (copy != null) { this.1 = null; }` since value could be different after the lock is acquired...?

Reply:
@franz1981 and this one?

Reply:
IDK, really...
the case in which this would happen can only be if:
- 2 removes are concurrently acquiring the lock after read a non-null instance: we do expect the second one to NOT perform any nulling out while within the lock, because the value has been already removed. Performing it again is a waste, but it isn't "wrong" let's say.
- 1 remove read a non-null instance, another remove cleanup everything and a compute acquire the lock, allocating a new instance: if that happen, the remove will enter and observe a different value from the first one, but is still fine if it read and remove it, I think. The problem is that it loose linearizability (but is due to the unguarded volatile load, as said in other comments).

In short, performing volatile reads always bring a new state in the mix, which makes all the operations non-atomic(s); we could do something must more "precise" here by creating an holder class `InstanceAndLock`, as said in other comments. This can make a much better choice here.

Has to be said, that, at least of request scoped beans we have an earlier atomic (based on compare and set) invalidation, which fully avoid remove to happen twice on the same context, which is the key requirement to make the latter case to be triggered.